File: georocket-server/src/test/java/io/georocket/index/xml/XalAddressIndexerTest.java
package io.georocket.index.xml;

import com.google.common.collect.ImmutableMap;
import io.georocket.util.XMLParserTransformer;
import io.vertx.core.buffer.Buffer;
import io.vertx.ext.unit.Async;
import io.vertx.ext.unit.TestContext;
import io.vertx.ext.unit.junit.VertxUnitRunner;
import org.junit.Test;
import org.junit.runner.RunWith;
import rx.Observable;

import java.io.IOException;
import java.io.InputStream;
import java.util.Map;
import java.util.Scanner;

/**
 * Tests {@link XalAddressIndexer}
 * @author Michel Kraemer
 */
@RunWith(VertxUnitRunner.class)
public class XalAddressIndexerTest {
  /**
   * Indexes the given XML file and checks if the result matches the
   * expected properties map
   * @param expected the expected properties map
   * @param xmlFile the XML file to parse
   * @param context the current test context
   * @throws IOException if the JSON file could not be read
   */
  private void assertIndexed(Map<String, Object> expected, String xmlFile,
      TestContext context) throws IOException {
    String json;
    try (InputStream is = getClass().getResourceAsStream(xmlFile);
         Scanner scanner = new Scanner(is, "UTF-8")) {
      scanner.useDelimiter("\\A");
      json = scanner.next();
    }

    XalAddressIndexer indexer = new XalAddressIndexer();
    Map<String, Object> expectedMap = ImmutableMap.of("address", expected);

    Async async = context.async();
    Observable.just(Buffer.buffer(json))
      .compose(new XMLParserTransformer())
      .doOnNext(indexer::onEvent)
      .last()
      .subscribe(r -> {
        context.assertEquals(expectedMap, indexer.getResult());
        async.complete();
      }, context::fail);
  }

  /**
   * Test if an XML file containing a XAL address can be indexed
   * @param context the current test context
   * @throws IOException if the XML file could not be read
   */
  @Test
  public void feature(TestContext context) throws IOException {
    Map<String, Object> expected = ImmutableMap.of(
      "Country", "Germany",
      "Locality", "Darmstadt",
      "Street", "Fraunhoferstra\u00DFe",
      "Number", "5"
    );
    assertIndexed(expected, "xal_simple_address.xml", context);
  }
}

File: georocket-server/src/test/kotlin/io/georocket/index/xml/XalAddressIndexerTest.kt
package io.georocket.index.xml

import io.georocket.coVerify
import io.vertx.core.Vertx
import io.vertx.core.buffer.Buffer
import io.vertx.junit5.VertxExtension
import io.vertx.junit5.VertxTestContext
import io.vertx.kotlin.coroutines.dispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith

/**
 * Tests [XalAddressIndexer]
 * @author Michel Kraemer
 */
@ExtendWith(VertxExtension::class)
class XalAddressIndexerTest {
  /**
   * Indexes the given XML file and checks if the result matches the
   * expected properties map
   */
  private suspend fun assertIndexed(expected: Map<String, Any>, xmlFile: String) {
    val json = javaClass.getResource(xmlFile)!!.readText()

    val indexer = XalAddressIndexer()
    val expectedMap = mapOf("address" to expected)

    XMLTransformer().transform(Buffer.buffer(json)).collect { e ->
      indexer.onEvent(e)
    }

    Assertions.assertThat(indexer.makeResult()).isEqualTo(expectedMap)
  }

  /**
   * Test if an XML file containing a XAL address can be indexed
   */
  @Test
  fun feature(ctx: VertxTestContext, vertx: Vertx) {
    CoroutineScope(vertx.dispatcher()).launch {
      val expected = mapOf(
        "Country" to "Germany",
        "Locality" to "Darmstadt",
        "Street" to "Fraunhoferstra\u00DFe",
        "Number" to "5"
      )

      ctx.coVerify {
        assertIndexed(expected, "xal_simple_address.xml")
      }

      ctx.completeNow()
    }
  }
}
